pkg/daemon/log_parsing.go

```diff
@@ -0,0 +1,167 @@
+	case ptp4lProcessName:
+		return parser.NewPTP4LExtractor()
+	case phc2sysProcessName:
+		return parser.NewPhc2SysExtractor()
+	case ts2phcProcessName:
+		return parser.NewTS2PHCExtractor()
+	default:
+		glog.Errorf("No parser available for process: %s", processName)
+		return nil
+	}
+}
+
+// processWithParser uses the new parser-based approach for processes with parsers
+func processWithParser(process *ptpProcess, output string) {
+	// Extract metrics and events using the parser
+	metrics, ptpEvent, err := process.logParser.Extract(output)
+	if err != nil {
+		glog.Errorf("Failed to extract metrics from %s output: %v", process.name, err)
+		return
+	}
+
+	process.hasCollectedMetrics = true
+
+	// Process metrics if available
+	if metrics != nil {
+		processParsedMetrics(process, metrics)
+	}
+
+	// Process PTP events if available
+	if ptpEvent != nil {
+		processParsedEvent(process, ptpEvent)
+	}
+}
+
+func processParsedMetrics(process *ptpProcess, ptpMetrics *parser.Metrics) {
+	// Convert interface from possible clock id
+	iface := process.ifaces.GetPhcID2IFace(ptpMetrics.Iface)
+	if iface != clockRealTime {
+		iface = utils.GetAlias(iface)
+	}
+
+	// Update PTP metrics using the parsed data
+	updatePTPMetrics(ptpMetrics.Source, process.name, iface, ptpMetrics.Offset, ptpMetrics.MaxOffset, ptpMetrics.FreqAdj, ptpMetrics.Delay)
+
+	// Update clock state metrics if available
+	if ptpMetrics.ClockState != "" {
+		updateClockStateMetrics(process.name, iface, string(ptpMetrics.ClockState))
+	}
+
+	configName := strings.Replace(strings.Replace(process.messageTag, "]", "", 1), "[", "", 1)
+	if configName != "" {
+		configName = strings.Split(configName, MessageTagSuffixSeperator)[0]
+	}
+
+	// Handle master offset source tracking
+	if ptpMetrics.Source == "master" && configName != "" {
+		masterOffsetSource.set(configName, process.name)
+	}
+
+	switch process.name {
+	case ptp4lProcessName:
+		if ptpMetrics.Iface != "" && configName != "" {
+			masterOffsetIface.set(configName, ptpMetrics.Iface)
+		}
+	case ts2phcProcessName:
+		// Send event for ts2phc
+		select {
+		case process.eventCh <- event.EventChannel{
+			ProcessName: event.TS2PHC,
+			State:       convertParserClockStateEventPTPState(ptpMetrics.ClockState),
+			CfgName:     configName,
+			IFace:       iface,
+			Values: map[event.ValueType]interface{}{
+				event.OFFSET:      int64(ptpMetrics.Offset),
+				event.NMEA_STATUS: int64(1),
+			},
+			ClockType:  process.clockType,
+			Time:       time.Now().UnixMilli(),
+			WriteToLog: true,
+			Reset:      false,
+		}:
+		default:
+		}
+	}
+}
+
+// processParsedEvent handles PTP events extracted by the parser
+func processParsedEvent(process *ptpProcess, ptpEvent *parser.PTPEvent) {
+	if process.name != ptp4lProcessName {
+		return
+	}
+
+	// Update interface role metrics
+	if ptpEvent.PortID > 0 && len(process.ifaces) >= ptpEvent.PortID-1 {
+		configName := strings.Replace(strings.Replace(process.messageTag, "]", "", 1), "[", "", 1)
```

Comment:
This should be len(process.ifaces) > ptpEvent.PortID-1
to prevent  any panic  and safe to access  process.ifaces[ptpEvent.PortID - 1]

Reply:
This is a combination of two if statements in the old code 
```
if portId, role := extractPTP4lEventState(output); portId > 0 {
	if len(ifaces) >= portId-1 {
```
https://github.com/k8snetworkplumbingwg/linuxptp-daemon/blob/main/pkg/daemon/metrics.go#L299-L300